_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,499 @@
+However, it misses one basic fact about these requests: Most of them 
+are made on behalf of the current user and not the requesting
+application. 
+
+So a posts controller like this:
+
+```ruby
+  module Api
+    module V1
+      class PostsController < ApplicationController
+        before_filter :restrict_access
+
+        def index
+          @posts = current_user.posts.page(1)
+          respond_with @posts
+        end
+
+        private
+
+        def restrict_access        
+          authenticate_or_request_with_http_token do |token, options|          
+            ApiKey.exists?(access_token: token)
+          end
+        end
+
+      end
+    end
+  end
+```
+
+Is, well, a little awkward. Because there's really no way to 
+know who the "current_user" is without passing in some extra variable.
+
+
+#### Enter OAuth.
+
+OAuth especially via [DoorKeeper](https://github.com/doorkeeper-gem/doorkeeper) 
+and [Omniauth](https://github.com/intridea/omniauth), is amazing, 
+but popping opening new windows (or mobile apps), redirect URIs, tokens, secrets, and the funky 
+user experience that often accompany them can be a bit too much
+ceremony especially if you're connecting two trusted applications.
+
+#### Tokens to the rescue!
+
+Encrypting strings that can be later decrypted (usually obfuscated to a word: Token)
+is a concept oft hidden by Wizard-Of-Oz style curtains within frameworks like Rails. 
+
+Just Set some random ```secret_base_key``` and you're good to go!
+
+When you dig in, oh boy, *Initialization Vectors*, *Ciphers*, wacky words / bad hacker nicknames, inspiring both dread and awe in the uninitiated. 
+But just like any good magic trick deciphering the slight of hand is both empowering and inspiring. 
+
+So let's take a peek behind the curtain.
+
+#### Imagine a scenario
+
+##### Application 1: 
+Our Authentication service (https://id.wework.io)
+In a more traditional architecture, this would be the OAuth provider.
+
+This service has a login API. For the sake of brevity, it uses [Devise](https://github.com/plataformatec/devise)
+to manage users and password. This application also utilizes CORS via [Rack CORs](https://github.com/cyu/rack-cors).
+Allowing appropriately configured cross-origin access.
+
+##### It has a user model
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+  # Devise password stuff
+  devise :database_authenticatable
+end
+```
+
+##### It has a base API controller
+
+```ruby
+module Api
+  class ApiController < ApplicationController  
+    skip_before_action :verify_authenticity_token
+  end
+end
+```
+
+##### It has a login controller endpoint:
+
+```ruby
+module Api
+  module V1
+    class LoginController < ApiController
+      
+      def create # POST: /api/v1/login
+        @user = User.find_by(email: params[:email])
+
+        if @user.present? && @user.valid_password?(params[:password])
```

Comment:
Maybe change to `if User.authenticate(params[:email], params[:password])?`
That way we don't give away implementation details